src/main/java/org/jvnet/hudson/test/JenkinsRule.java

```diff
@@ -1265,28 +1269,89 @@ public Page search(String q) throws Exception {
-            throw new IllegalStateException("Unexpected SAXException.", e);
-        }
-        WebResponse webResponse = runsPage.getWebResponse();
+        return getJSON(path, webClient);
+    }
+
+    /**
+     * Get JSON from a Jenkins relative endpoint.
+     * You can preconfigure webClient for example to set a token for authentication, or accept error HTTP status.
+     * @param path relative path, should not start with '/'
+     * @param webClient a pre-configured web client.
+     * @return The JSON response from server.
+     * @throws IOException
+     */
+    public JSONWebResponse getJSON(@NonNull String path, @NonNull JenkinsRule.WebClient webClient) throws IOException {
+        assert !path.startsWith("/");
+
+        URL URLtoCall = new URL(getURL(), path);
+        WebRequest getRequest = new WebRequest(URLtoCall, HttpMethod.GET);
+        getRequest.setAdditionalHeader("Content-Type","application/json");
+        getRequest.setAdditionalHeader("Accept", "application/json");
+        getRequest.setAdditionalHeader("Accept-Encoding", "*");
+
+        return new JSONWebResponse(webClient.loadWebResponse(getRequest));
+    }
+
+    /**
+     * Send JSON content to a Jenkins relative endpoint.
+     * You can preconfigure webClient for example to set a token for authentication, or accept error HTTP status.
+     * @param path relative path, should not start with '/'
+     * @param json the json payload to send
+     * @param webClient a pre-configured web client.
+     * @return The JSON response from server.
+     * @throws IOException
+     */
+    public JSONWebResponse putJSON(@NonNull String path, @NonNull JSON json, @NonNull JenkinsRule.WebClient webClient) throws IOException {
+        assert !path.startsWith("/");
+
+        URL URLtoCall = new URL(getURL(),path);
+        WebRequest putRequest = new WebRequest(URLtoCall, HttpMethod.PUT);
+        putRequest.setRequestBody(json.toString());
+        putRequest.setAdditionalHeader("Content-Type","application/json");
+        putRequest.setAdditionalHeader("Accept", "application/json");
+        putRequest.setAdditionalHeader("Accept-Encoding", "*");
+
+        return new JSONWebResponse(webClient.loadWebResponse(putRequest));
+    }
+
+    /**
+     * POST JSON content to a Jenkins relative endpoint using the user token for authentication.
+     * You can preconfigure webClient for example to accept error HTTP status.
+     *
+     * @param path relative path, should not start with '/'
+     * @param json the json payload to send
+     * @param webClient a pre-configured web client either with crumb or token.
+     * @param user user to use for authentication
+     * @return The JSON response from server.
+     * @throws IOException
+     */
+    public JSONWebResponse postJSON(@NonNull String path, @NonNull JSON json, @NonNull WebClient webClient,
+                                    @NonNull User user) throws IOException {
```

Comment:
Why not requesting a pre-configured webClient, and ask for the caller to do `webClient.withBasicApiToken(user)` ?

Reply:
because it will not work at all if the client is not pre-configured with withBasicApiToken() ; when for get it's working fine without token. So I found it misleading to just pass a webClient in this case, since having a User is mandatory.

Reply:
And I tried to fallback on crumb if there is no user, it's complicated to do so, because you need to check if a user token has been added and there is no existing method to tell "is the token set ?". Since the other postJSON works fine with crumb, I think this one should be used when authentication is needed, and the other method when you don't want to setup authentication on your test.

Reply:
and also if the crumb is set, then you cannot add a token anymore.

Reply:
> having a User is mandatory

In a realistic installation perhaps, but plenty of functional tests do not bother setting up authentication & authorization, and just want to verify behavior of an endpoint unrelated to permission checks.

Reply:
Now I see your point about a crumb being demanded in the case of an anonymous request, unless of course there is a `CrumbExclusion`. Still, clearer to have the test call one of the `withBasicApiToken` overloads.

Reply:
code moved